storage/plans/iscsi/iscsi.go

```diff
@@ -0,0 +1,390 @@
+		return nil, errors.Errorf("missing required field: address")
+	}
+	if port, ok = info["port"]; !ok {
+		return nil, errors.Errorf("missing required field: port")
+	}
+
+	iPort, err := strconv.Atoi(port)
+	if err != nil {
+		return nil, errors.Errorf("invalid port: %v", port)
+	}
+	user, _ = info["chap-user"]
+	secret, _ = info["chap-secret"]
+	plan := &iscsiConnectionInfo{
+		iqn:        iqn,
+		address:    address,
+		port:       iPort,
+		chapSecret: secret,
+		chapUser:   user,
+	}
+	return plan, nil
+}
+
+// sessionBase returns the iSCSI sysfs session folder
+func (i *iscsiConnectionInfo) sessionBase(deviceName string) (string, error) {
+	lnkPath := filepath.Join(sysfsBlock, deviceName)
+	lnkRealPath, err := os.Readlink(lnkPath)
+	if err != nil {
+		return "", err
+	}
+	fullPath, err := filepath.Abs(filepath.Join(sysfsBlock, lnkRealPath))
+	if err != nil {
+		return "", err
+	}
+	segments := strings.SplitN(fullPath[1:], "/", -1)
+	if len(segments) != 9 {
+		// iscsi block devices look like:
+		// /sys/devices/platform/host2/session1/target2:0:0/2:0:0:1/block/sda
+		return "", errors.Errorf("not an iscsi device")
+	}
+	if _, err := os.Stat(filepath.Join(sysfsiSCSIHost, segments[3])); err != nil {
+		return "", errors.Errorf("not an iscsi device")
+	}
+	sessionPath := filepath.Join(sysfsiSCSISession, segments[4])
+	if _, err := os.Stat(sessionPath); err != nil {
+		return "", errors.Errorf("session does not exits")
+	}
+	return sessionPath, nil
+}
+
+func (i *iscsiConnectionInfo) deviceName() (string, error) {
+	items, err := ioutil.ReadDir(sysfsBlock)
+	if err != nil {
+		return "", err
+	}
+
+	for _, val := range items {
+		sessionBase, err := i.sessionBase(val.Name())
+		if err != nil {
+			logger.Tracef("failed to get session folder for device %s: %s", val.Name(), err)
+			continue
+		}
+		tgtnameFile := filepath.Join(sessionBase, "targetname")
+		if _, err := os.Stat(tgtnameFile); err != nil {
+			logger.Tracef("%s was not found. Skipping", tgtnameFile)
+			continue
+		}
+		tgtname, err := ioutil.ReadFile(tgtnameFile)
+		if err != nil {
+			return "", err
+		}
+		trimmed := strings.TrimSuffix(string(tgtname), "\n")
+		if trimmed == i.iqn {
+			return val.Name(), nil
+		}
+	}
+	return "", errors.NotFoundf("device for iqn %s not found", i.iqn)
+}
+
+func (i *iscsiConnectionInfo) portal() string {
+	return fmt.Sprintf("%s:%d", i.address, i.port)
+}
+
+func (i *iscsiConnectionInfo) configFile() string {
+	hostPortPair := fmt.Sprintf("%s,%d", i.address, i.port)
+	return filepath.Join(iscsiConfigFolder, i.iqn, hostPortPair)
+}
+
+func (i *iscsiConnectionInfo) isNodeConfigured() bool {
+	if _, err := os.Stat(i.configFile()); err != nil {
+		return false
+	}
+	return true
+}
+
+// addTarget adds the iscsi target config
+func (i *iscsiConnectionInfo) addTarget() error {
+	newNodeParams := []string{
+		"iscsiadm", "-m", "node",
+		"-o", "new",
+		"-T", i.iqn,
+		"-p", i.portal()}
+	result, err := runCommand(newNodeParams)
```

Comment:
Given that these take mostly the same parameters, I wonder if it is worth having a helper function.
That said, it can also be nice to see the whole command that is being executed in one spot, rather than having to look around for which parameters are/aren't being passed.
Probably the clarity of what full command is being run wins out in the end.
They seem to all take -m, -o, -T, but not all take -p/-n/-v. Would it be clearer as a parameter struct?

runISCSICommand(iscsiParameter{
  Mode: "node",
  Operation: "update",
  T?: i.iqn,
  N?: "node.startup",
  V?: "automatic",
  Portal: i.portal(),
etc.

Think about it.

Reply:
So, the whole module is around 400 lines of code, which essentially does a simple thing like running a bunch of exec commands. Module is pretty self contained and simple. I would like to avoid over engineering it. I avoided creating a struct, because the task seemed simple enough and clear enough to get away with a simple ```[]string{}```. Anything more complicated than that seemed overkill for the current purpose.